janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.transaction.vertexcache;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.google.common.base.Preconditions;
+import org.jctools.maps.NonBlockingHashMapLong;
+import org.janusgraph.graphdb.internal.InternalVertex;
+import org.janusgraph.graphdb.vertices.AbstractVertex;
+import org.janusgraph.util.datastructures.Retriever;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+public class CaffeineVertexCache implements VertexCache {
+    private static final Logger log =
+        LoggerFactory.getLogger(CaffeineVertexCache.class);
+
+    private final ConcurrentMap<Long, InternalVertex> volatileVertices;
```

Comment:
Caffeine has special feature for avoid this extra field - https://github.com/ben-manes/caffeine/issues/264

Implementation could be looks like https://gist.github.com/mad/9119f48812000a03dea16c7485c621fc

@ministat @porunov

Reply:
> Caffeine has special feature for avoid this extra field - [ben-manes/caffeine#264](https://github.com/ben-manes/caffeine/issues/264)
> 
> Implementation could be looks like https://gist.github.com/mad/9119f48812000a03dea16c7485c621fc
> 
> @ministat @porunov

That's interesting! I would prefer refactoring this class to eliminate this extra field as you proposed. That said, as it is a performance related part of the code we should probably execute benhmarking for any improvement we make here, just to make sure we don't have a regression. Other than that your proposal is much cleaner, so that's great.
Should we open a separate ticket to track this?

Reply:
Also, a related transitioning PR which replaces other Guava caches with Caffeine cache is: #3198 
After that transitioning is merged I believe we need to refactor [CaffeineVertexCache.java](https://github.com/JanusGraph/janusgraph/blob/c30c7071227de338a63d82336b34f749f2384b22/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java) and [ExpirationKCVSCache.java](https://github.com/JanusGraph/janusgraph/blob/0822cce3d7a50cdb530d13a363f076b18f9b14be/janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/cache/ExpirationKCVSCache.java).